### PR TITLE
Support ExternalDNS integration by DNSEndpoint

### LINKED ISSUE #88: Automate associating custom domain name with service
Today,  customer needs manually configure route53, add a CNAME record to map custom domain name to a VPC lattice generated DNS name

Ideally,  lattice controller can automate this process .

### PR DESCRIPTION
**Pull request by solmonk**
**What type of PR is this?**

feature

**Which issue does this PR fix**:

#88

**What does this PR do / Why do we need it**:

Resolves #88 by using DNSEndpoint CRD supported by ExternalDNS

**If an issue # is not available please add repro steps and logs from aws-gateway-controller showing the issue**:

**Testing done on this change**:

`make e2etest` and manual traffic tests

`HTTPRoute` resource with hostname will create the resource resembles the following. Note the `ownerReferences` field.

```yaml
apiVersion: externaldns.k8s.io/v1alpha1
kind: DNSEndpoint
metadata:
  creationTimestamp: "2023-07-10T23:35:29Z"
  generation: 1
  name: byoc-parking-dns
  namespace: default
  ownerReferences:
  - apiVersion: gateway.networking.k8s.io/v1beta1
    blockOwnerDeletion: true
    controller: true
    kind: HTTPRoute
    name: byoc-parking
    uid: 4f494890-8314-4a37-b787-99fe80cedd0a
  resourceVersion: "17044921"
  uid: ea7f0852-9060-4927-9d89-b5f4477acdfe
spec:
  endpoints:
  - dnsName: parking.my-test.com
    recordTTL: 300
    recordType: CNAME
    targets:
    - byoc-parking-default-03f7a7580d37de850.7d67968.vpc-lattice-svcs.us-west-2.on.aws
```

**Automation added to e2e**:

**Will this PR introduce any new dependencies?**:

**Will this break upgrades or downgrades. Has updating a running cluster been tested?**:

**Does this PR introduce any user-facing change?**:

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by liwenwu-amazon** [comment]
are you planning to add e2e test for this as part of this PR? Thanks

**Comment by liwenwu-amazon** [comment]
QQ, for customer who do *NOT* run external DNS, will they able to use lattice controller at all?

**Review by liwenwu-amazon — CHANGES_REQUESTED** [review]
I think we should `not` require customer to "MUST" run external DNS.   The lattice controller should still run if they choose to manually or another way to update DNS.

**Comment by solmonk** [comment]
Running ExternalDNS is not required at all (and should not be). I have a couple of setups in this PR for it:
* Watch for DNSEndpoint will be disabled when CRD is not detected - see controller setup
* DNSEndpoint creation does not happen when CRD is not detected. This is done by detecting `NoMatchError` on listing DNSEndpoint resource.
* The documentation mentions that customers can setup DNS manually when not using ExternalDNS
* I have done the testing with deleting DNSEndpoint CRD

**Comment by liwenwu-amazon** [comment]
> 

Please add some test for these business logic. thanks

File: controllers/httproute_controller.go [hunk 1]
```diff
@@ -336,11 +339,41 @@ func (r *HTTPRouteReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
 		Watches(&source.Kind{Type: &gateway_api.Gateway{}}, gwEventHandler).
 		Watches(&source.Kind{Type: &corev1.Service{}}, svcEventHandler).
-		Watches(&source.Kind{Type: &mcs_api.ServiceImport{}}, svcImportEventHandler).
-		Complete(r)
+		Watches(&source.Kind{Type: &mcs_api.ServiceImport{}}, svcImportEventHandler)
+
+	if ok, err := isExternalDNSSupported(mgr); ok {
+		builder.Owns(&endpoint.DNSEndpoint{})
+	} else if err != nil {
+		glog.V(2).Infof("Unknown error while discovering CRD: %v", err)
```
Comment (liwenwu-amazon) [inline controllers/httproute_controller.go:353]: This is concerning.  Looks like if ExternalDNS is not supported, the controller will `not` run.

Comment (solmonk) [reply controllers/httproute_controller.go:353]: `err != nil` just means that it failed to run discovery API for some reason. If ExternalDNS isn't supported the result is `(false, nil)` so it gracefully skips the logic.

**Comment by solmonk** [comment]
Will try to update a few more unit test cases. I am hoping to add e2e test case in other PR. There is a separate ongoing work on E2E testing for custom domain names so I am worried about the conflict. I can add more check when it is delivered.

File: pkg/deploy/lattice/service_synthesizer.go [hunk 2]
```diff
@@ -65,6 +70,14 @@ func (s *serviceSynthesizer) Synthesize(ctx context.Context) error {
 			s.latticeDataStore.AddLatticeService(resService.Spec.Name, resService.Spec.Namespace,
 				serviceStatus.ServiceARN, serviceStatus.ServiceID, serviceStatus.ServiceDNS)
 
+			glog.V(6).Infof("Creating Service DNS")
+			resService.Status = &serviceStatus
+			err = s.dnsEndpointManager.Create(ctx, resService)
+			if err != nil {
+				glog.V(6).Infof("Failed creating service DNS: %v", err)
```
Comment (liwenwu-amazon) [inline pkg/deploy/lattice/service_synthesizer.go:77]: can u make this V(2) for debugging purpose?

File: pkg/deploy/lattice/service_synthesizer_test.go [hunk 3]
```diff
@@ -161,7 +163,11 @@ func Test_SynthesizeService(t *testing.T) {
 			mockSvcManager.EXPECT().Delete(ctx, latticeService).Return(tt.mgrErr)
 		}
 
-		synthesizer := NewServiceSynthesizer(mockSvcManager, stack, ds)
+		if !spec.IsDeleted && tt.mgrErr == nil {
```
Comment (liwenwu-amazon) [inline pkg/deploy/lattice/service_synthesizer_test.go:166]: can u also add negative test?

File: controllers/httproute_controller.go (hunk 1 above)
Comment (liwenwu-amazon) [reply controllers/httproute_controller.go:353]: can u add this to the comment? it was not obvious to me when I read this code. thanks

**Comment by liwenwu-amazon** [comment]
> Will try to update a few more unit test cases. I am hoping to add e2e test case in other PR. There is a separate ongoing work on E2E testing for custom domain names so I am worried about the conflict. I can add more check when it is delivered.

QQ, have you run e2e 
* have external DNS running, but no DNS  CRD?
* do not have external DNS and also no CRD?
thanks

**Comment by solmonk** [comment]
Whether externalDNS is running or not does not matter to our controller. We could perhaps just check with CRD / no CRD scenario.

File: docs/developer.md [hunk 4]
```diff
@@ -16,6 +16,7 @@ make presubmit
 kubectl apply -f config/crds/bases/k8s-gateway-v0.6.1.yaml
 kubectl apply -f config/crds/bases/multicluster.x-k8s.io_serviceexports.yaml
 kubectl apply -f config/crds/bases/multicluster.x-k8s.io_serviceimports.yaml
+kubectl apply -f config/crds/bases/externaldns.k8s.io_dnsendpoints.yaml
```
Comment (liwenwu-amazon) [inline docs/developer.md:19]: Please also update "End-toEnd" testing section,  that e2e should run `OK` with both
* with DNS CRD
* `without`DNS CRD

File: pkg/deploy/lattice/service_synthesizer.go (hunk 2 above)
Comment (solmonk) [reply pkg/deploy/lattice/service_synthesizer.go]: fixed

File: pkg/deploy/lattice/service_synthesizer_test.go (hunk 3 above)
Comment (solmonk) [reply pkg/deploy/lattice/service_synthesizer_test.go:190]: fixed

File: controllers/httproute_controller.go (hunk 1 above)
Comment (solmonk) [reply controllers/httproute_controller.go]: fixed

File: docs/developer.md (hunk 4 above)
Comment (solmonk) [reply docs/developer.md:19]: Actually I have bundled DNSEndpoint CRD with both helm chart and `deploy.yaml`, so it is just like other CRDs that we assume its existence. Though I have added missing CRD handling to assist local development, I don't think we have to force the setting for e2etest.

**Comment by zijun726911** [comment]
When people use ExternalDNS, do they usually create `kind:DNSEndpoint` by themselves, or it has some use cases that some controllers could create `kind:DNSEndpoint` on behalf of users?

**Comment by solmonk** [comment]
@zijun726911 Both are feasible scenarios, for automatic management I think [this project](https://github.com/redhat-cop/global-load-balancer-operator) is a good example.

File: pkg/deploy/lattice/service_synthesizer.go [hunk 5]
```diff
@@ -65,6 +70,14 @@ func (s *serviceSynthesizer) Synthesize(ctx context.Context) error {
...
 				serviceStatus.ServiceARN, serviceStatus.ServiceID, serviceStatus.ServiceDNS)
 
+			glog.V(6).Infof("Creating Service DNS")
+			resService.Status = &serviceStatus
+			err = s.dnsEndpointManager.Create(ctx, resService)
+			if err != nil {
+				glog.V(2).Infof("Failed creating service DNS: %v", err)
+				return err
+			}
+
```
Comment (zijun726911) [inline pkg/deploy/lattice/service_synthesizer.go:80]: Do we need to do`s.latticeDataStore.AddLatticeService()` at last, after  s.dnsEndpointManager.Create() finished? i.e., if `s.dnsEndpointManager.Create()` failed we should not do `s.latticeDataStore.AddLatticeService()`?

**Comment by zijun726911** [comment]
Can you describe more manual e2etest result in the PR description? for example, if you installed the CoreDNS addon in your cluster and create a httproute with hostnames it indeed can create a route53 CNAME record

File: pkg/deploy/lattice/service_synthesizer.go (hunk 5 above)
Comment (solmonk) [reply pkg/deploy/lattice/service_synthesizer.go:80]: dnsEndpoint is not a requirement for a service to function so I think we can still unblock other logic that uses it. I think latticeDataStore should represent what controller actually has in Lattice.

File: docs/developer.md (hunk 4 above)
Comment (liwenwu-amazon) [reply docs/developer.md:19]: Customer may choose to have a different way (other than the deploy.yaml or helm in github) to run lattice controller.  If we do not enforce testing it e2e and  miss a regression, it  can impact these customers

Comment (solmonk) [reply docs/developer.md:19]: Updated the doc

**Review by liwenwu-amazon — APPROVED** [review]
LGTM! Thank you

**Pull request merged**